Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerHandler.swift

```diff
@@ -0,0 +1,564 @@
+/*
+ * Copyright 2021, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#if compiler(>=5.5)
+
+import _NIOConcurrency
+import NIOCore
+import NIOHPACK
+
+@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+public struct GRPCAsyncServerHandler<
+  Serializer: MessageSerializer,
+  Deserializer: MessageDeserializer
+>: GRPCServerHandlerProtocol {
+  @usableFromInline
+  internal let _handler: _GRPCAsyncServerHandler<Serializer, Deserializer>
+
+  public func receiveMetadata(_ metadata: HPACKHeaders) {
+    self._handler.receiveMetadata(metadata)
+  }
+
+  public func receiveMessage(_ bytes: ByteBuffer) {
+    self._handler.receiveMessage(bytes)
+  }
+
+  public func receiveEnd() {
+    self._handler.receiveEnd()
+  }
+
+  public func receiveError(_ error: Error) {
+    self._handler.receiveError(error)
+  }
+
+  public func finish() {
+    self._handler.finish()
+  }
+}
+
+@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+extension GRPCAsyncServerHandler {
+  public typealias Request = Deserializer.Output
+  public typealias Response = Serializer.Input
+
+  @inlinable
+  public init(
+    context: CallHandlerContext,
+    requestDeserializer: Deserializer,
+    responseSerializer: Serializer,
+    interceptors: [ServerInterceptor<Request, Response>],
+    wrapping unary: @escaping @Sendable(Request, GRPCAsyncServerCallContext) async throws
+      -> Response
+  ) {
+    self._handler = .init(
+      context: context,
+      requestDeserializer: requestDeserializer,
+      responseSerializer: responseSerializer,
+      interceptors: interceptors,
+      userHandler: { requestStream, responseStreamWriter, context in
+        guard let request = try await requestStream.prefix(1).first(where: { _ in true }) else {
```

Comment:
I thought I asked elsewhere but I couldn't find the comment: do we need both `prefix` and `first`?

Reply:
Yes, I think we should. The only harm in doing this is that we have to wait for the end of the request stream before we can invoke the function which is potentially slower (as the end might be delivered in a separate http/2 frame to the request).

Reply:
Great. I've updated this logic in d8f645e2.